src/utils/jsonable.js

```diff
@@ -37,3 +45,56 @@ export type JSONable =
 //
 // ... which is consistent with the definition of inexact types, if
 // inconvenient.
+
+/**
+ * Approximate type of a JSONable value received as input.
+ *
+ * Unlike the definition of `JSONableDict`, this permits the possibility of
+ * `void` elements as dictionary members. This is a workaround for Flow's
+ * by-design soundness hole involving indexer elements:
+ *
+ * ```flow
+ *   const obj: { [string]: string } = { a: "a" };
+ *   const value: string = obj['b'];
+ *   value.trim();  // kaboom!
+ * ```
+ *
+ * ... that is, Flow will assume **all subscript accesses to indexer-property-
+ * typed values are valid**, and will not mark values extracted therefrom as
+ * potentially undefined. (For references indicating that this is, indeed, by
+ * design, see [0] and [1].)
+ *
+ * Adding `void` as a possible dictionary entry type to JSONable would solve
+ * this problem -- but only at the cost of introducing others:
+ *   * For output, it is possible to feed a `void`-containing structure to a
+ *     serializing function, even though that structure cannot be reversibly
+ *     JSON- serialized.
+ *   * When processing input, `Object.values` (and related functions) will yield
+ *     a type which contains `void`, even though no value in a just-parsed JSON
+ *     object can actually be `undefined`.
+ *
+ * We can mostly avoid these problems by creating a secondary JSONable-like
+ * type, `JSONableInput`, used only for validation and transformation of JSON
+ * input data.
+ *
+ * As JSONableInput <: JSONable, JSON values should generally be passed around
+ * as `JSONable` until and unless they're actually being validated; they can
+ * then be locally downgraded to `JSONableInput`.
```

Comment:
Rather JSONable <: JSONableInput, right? I read that operator as "is a subtype of".

And then... "downgraded" sure is an accurate characterization (as it's e.g. a less informative type), but potentially a confusing word given the mixed metaphor with the type hierarchy :upside_down_face: . Maybe just "be locally cast to"? Or s/cast/upcast/.

Reply:
An upcast is a downgrade. There's nothing confusing there.  ⍩⃝  

Fixed and amended.